Controllers as services

derrabus opened the pull request:
This PR removes the `ContainerAware` trait from all controllers, thus making them independant from the Silex application. This refactoring can be a preparation for migrating away from Silex.

Related to #618.

localheinz commented on classes/Http/Controller/Admin/SpeakersController.php:
```diff
@@ -23,30 +23,55 @@
+    private $capsule;
+    private $applicationArrival;
+    private $applicationDeparture;
+    private $applicationAirport;
```
Can we add docblocks here?

derrabus replied:
Sure, I'll add doc blocks to all newly introduced fields.

derrabus replied:
To explain myself: I usually leave them out for dependencies, because the type-hints on the constructor already document the types of the fields.

localheinz commented on classes/Http/Controller/Admin/TalksController.php:
```diff
@@ -19,26 +19,40 @@
 {
+    private $authentication;
+    private $talkFilter;
+    private $talkHandler;
```
Can we add docblocks here?

localheinz commented on classes/Http/Controller/Admin/TalksController.php:
```diff
@@ -145,22 +145,16 @@ public function commentCreateAction(Request $request)
+                'type'  => 'success',
+                'short' => 'Success',
+                'ext'   => 'Comment Added!',
+            ]);
```
Can we adjust the indentation here?

localheinz commented on classes/Http/Controller/Admin/TalksController.php:
```diff
@@ -145,22 +145,16 @@ public function commentCreateAction(Request $request)
+            ]);
 
-        return $this->app->redirect($this->url('admin_talk_view', ['id' => $talkId]));
+        return new RedirectResponse($this->url('admin_talk_view', ['id' => $talkId]));
```
👍

localheinz commented on classes/Http/Controller/BaseController.php:
```diff
@@ -18,19 +18,24 @@
 {
-    use ContainerAware;
+    protected $twig;
+    protected $urlGenerator;
```
Can we add docblocks here?

localheinz commented on classes/Http/Controller/BaseController.php:
```diff
@@ -93,27 +92,20 @@ protected function render($name, array $context = [], $status = Response::HTTP_O
-        $request = $this->service('request_stack')->getCurrentRequest();
-
-        return $this->app->redirect($request->headers->get('referer'));
+        return new RedirectResponse($request->headers->get('referer'));
```
👍

localheinz commented on classes/Http/Controller/DashboardController.php:
```diff
@@ -14,33 +14,29 @@
-         * @var Speakers $speakers
-         */
-        $speakers = $this->service('application.speakers');
+    private $speakers;
```
Can we add a docblock here?

localheinz commented on classes/Http/Controller/ForgotController.php:
```diff
@@ -17,19 +17,34 @@
 use OpenCFP\Http\Form\ForgotForm;
 use OpenCFP\Http\Form\ResetForm;
-use Symfony\Component\Form\FormFactory;
+use Symfony\Component\Form\FormFactoryInterface;
```
👍

localheinz commented on classes/Http/Controller/ForgotController.php:
```diff
@@ -17,19 +17,34 @@
 {
+    private $formFactory;
+    private $accounts;
+    private $resetEmailer;
```
Can we add docblocks here?

localheinz commented on classes/Http/Controller/ProfileController.php:
```diff
@@ -13,27 +13,41 @@
 {
+    private $authentication;
+    private $purifier;
+    private $profileImageProcessor;
```
Can we add docblocks here?

localheinz commented on classes/Http/Controller/ProfileController.php:
```diff
@@ -92,37 +100,31 @@ public function processAction(Request $request)
+                'type'  => 'error',
+                'short' => 'Error',
+                'ext'   => \implode('<br>', $form->getErrorMessages()),
+            ]);
```
Can we adjust the indentation here?

localheinz commented on classes/Http/Controller/ProfileController.php:
```diff
@@ -92,37 +100,31 @@ public function processAction(Request $request)
-
-                $sanitizedData['photo_path'] = $profileImageProcessor->process($formData['speaker_photo']);
+                $sanitizedData['photo_path'] = $this->profileImageProcessor
+                    ->process($formData['speaker_photo']);
```
Not sure about the wrapping here, but I don't mind.

derrabus replied:
Sorry, I guess my IDE did that automatically.

localheinz commented on classes/Http/Controller/ProfileController.php:
```diff
@@ -147,18 +146,11 @@ public function passwordProcessAction(Request $request)
         $form->sanitize();
 
-        /** @var Session\Session $session */
-        $session = $this->service('session');
```
Since we would otherwise be invoking `$request->getSession()` multiple times, can we maybe just use `$session` here and assign it `$request->getSession()`?

derrabus replied:
See the discussion above.

localheinz commented on classes/Http/Controller/ForgotController.php:
```diff
@@ -40,20 +55,14 @@ public function indexAction()
         $form->handleRequest($request);
 
-        /** @var Session\Session $session */
-        $session = $this->service('session');
```
Since we would otherwise be invoking `$request->getSession()` multiple times, can we maybe just use `$session` here and assign it `$request->getSession()`?

derrabus replied:
See the discussion above.

derrabus replied:
See the discussion above.

localheinz commented on classes/Http/Controller/Admin/SpeakersController.php:
```diff
@@ -165,20 +178,11 @@ public function deleteAction(Request $request)
-        $authentication = $this->service(Authentication::class);
-
-        /** @var Session\Session $session */
-        $session = $this->service('session');
```
Since we would otherwise be invoking `$request->getSession()` multiple times, can we maybe just use `$session` here and assign it `$request->getSession()`?

derrabus replied:
`$request->getSession()` isn't an expensive operation, but yes, I can change that.

derrabus replied:
On the other hand: `getSession()` is usually followed by a `return` statement, so we're not really invoking `getSession()` multiple times. I wouldn't change it, tbh.

localheinz commented on classes/Http/Controller/Admin/SpeakersController.php:
```diff
@@ -209,20 +213,14 @@ public function demoteAction(Request $request)
         $id   = (int) $request->get('id');
 
-        /** @var Session\Session $session */
-        $session = $this->service('session');
```
Since we would otherwise be invoking `$request->getSession()` multiple times, can we maybe just use `$session` here and assign it `$request->getSession()`?

derrabus replied:
See the discussion above.

derrabus replied:
See the discussion above.

localheinz commented on classes/Http/Controller/ForgotController.php:
```diff
@@ -121,10 +120,7 @@ public function resetAction(Request $request)
 
         if ($error > 0) {
-            /** @var Session\Session $session */
-            $session = $this->service('session');
```
Since we would otherwise be invoking `$request->getSession()` multiple times, can we maybe just use `$session` here and assign it `$request->getSession()` (we probably need to pull it out of the condition)?

derrabus replied:
See the discussion above.

localheinz commented on classes/Http/Controller/ForgotController.php:
```diff
@@ -224,24 +202,18 @@ public function updatePasswordAction(Request $request)
         }
 
-        /** @var Session\Session $session */
-        $session = $this->service('session');
```
Since we would otherwise be invoking `$request->getSession()` multiple times, can we maybe just use `$session` here and assign it `$request->getSession()`?

derrabus replied:
See the discussion above.

localheinz commented on classes/Http/Controller/ProfileController.php:
```diff
@@ -68,16 +82,10 @@ public function editAction(Request $request)
-        $userId = $authentication->userId();
-
-        /** @var Session\Session $session */
-        $session = $this->service('session');
```
Since we would otherwise be invoking `$request->getSession()` multiple times, can we maybe just use `$session` here and assign it `$request->getSession()`?

derrabus replied:
See the discussion above.

localheinz commented on classes/Http/Controller/Reviewer/DashboardController.php:
```diff
@@ -20,16 +20,23 @@
 class DashboardController extends BaseController
 {
-    public function indexAction()
+    private $authentication;
```
Can we add a docblock here?

localheinz commented on classes/Http/Controller/Reviewer/SpeakersController.php:
```diff
@@ -18,10 +18,20 @@
 
 class SpeakersController extends BaseController
 {
+    private $reviewerUsers;
```
Can we add a docblock here?

localheinz commented on classes/Http/Controller/Reviewer/TalksController.php:
```diff
@@ -20,26 +20,39 @@
 {
+    private $authentication;
+    private $talkFilter;
+    private $talkHandler;
```
Can we add docblocks here?

localheinz commented on classes/Http/Controller/SecurityController.php:
```diff
@@ -16,10 +16,23 @@
 
 class SecurityController extends BaseController
 {
+    private $authentication;
```
Can we add a docblock here?

localheinz commented on classes/Http/Controller/SignupController.php:
```diff
@@ -18,27 +18,37 @@
-        $auth = $this->service(Authentication::class);
+    private $authentication;
+    private $accounts;
+    private $callForPapers;
```
Can we add docblocks here?

localheinz commented on classes/Http/Controller/SignupController.php:
```diff
@@ -53,49 +63,43 @@ public function indexAction()
+            return $this->redirectBack($request);
         } catch (\RuntimeException $e) {
-            $this->app['session']->set('flash', [
+            $request->getSession()->set('flash', [
```
Since we would otherwise be invoking `$request->getSession()` multiple times, can we maybe just extract a  `$session` variable here?

localheinz commented on classes/Http/Controller/TalkController.php:
```diff
@@ -15,415 +15,38 @@
-    {
-        /** @var TalkHelper $helper */
-        $helper = $this->service(TalkHelper::class);
+    private $speakers;
```
Can we add a docblock here?

localheinz commented on classes/Http/Controller/TalkController.php:
```diff
@@ -15,415 +15,38 @@
      * @param Request $request
      *
-     * @return mixed
+     * @return Response
```
👍

localheinz commented on classes/Http/Controller/TalkController.php:
```diff
@@ -15,415 +15,38 @@
         return $this->render('talk/view.twig', \compact('talkId', 'talk'));
     }
-
-    public function editAction(Request $request)
```
~~Not sure what is happening here, has this functionality moved elsewhere?~~

Ah, I see, the `TalkFormController` below!

@BackEndTea @chartjes @mdwheele 

What do you think about extracting *actions* sooner or later, as functors?

This would have the following advantage

* dependencies are limited to what is needed for a single action (right now we would inject dependencies required for all actions, even if not all of them are used for a single request)
* routing definitions are more transparent (as we can ideally just use the class name, no more hard-coded class names concatenated with `:` and the action)

derrabus replied:
Yes, I've split the controller because we would inject too many unnecessary dependencies otherwise. And yes, I'd prefer single-action controllers as well, but that would've been out of scope for this PR.

localheinz replied:
Absolutely, @derrabus! Happy to start dissecting! 👍

chartjes replied:
I too think this would be a great idea once we get this PR merged.

localheinz commented on classes/Http/Controller/TalkFormController.php:
```diff
@@ -0,0 +1,149 @@
+{
+    private $authentication;
+    private $talkHelper;
+    private $callForPapers;
```
Can we add docblocks here?

localheinz commented on classes/Http/Controller/TalkFormController.php:
```diff
@@ -0,0 +1,149 @@
+            return $this->redirectTo('dashboard');
+        }
+
+        $data = [
```
While we are at it, can we inline `$data`?

Also see #851, where I already started working on that.

localheinz commented on classes/Http/Controller/TalkFormController.php:
```diff
@@ -0,0 +1,149 @@
+                [
+                    'type'  => 'error',
+                    'short' => 'Read Only',
+                    'ext'   => 'You cannot edit talks once the call for papers has ended', ]
```
What do you think about adjusting the wrapping here to

```diff
diff --git a/classes/Http/Controller/TalkFormController.php b/classes/Http/Controller/TalkFormController.php
index bb4cba0d..51b44ac9 100644
--- a/classes/Http/Controller/TalkFormController.php
+++ b/classes/Http/Controller/TalkFormController.php
@@ -54,7 +54,8 @@ class TalkFormController extends BaseController
                 [
                     'type'  => 'error',
                     'short' => 'Read Only',
-                    'ext'   => 'You cannot edit talks once the call for papers has ended', ]
+                    'ext'   => 'You cannot edit talks once the call for papers has ended',
+                ]
             );

             return new RedirectResponse($this->url('talk_view', ['id' => $talkId]));
```

or

```diff
diff --git a/classes/Http/Controller/TalkFormController.php b/classes/Http/Controller/TalkFormController.php
index bb4cba0d..ad88ffed 100644
--- a/classes/Http/Controller/TalkFormController.php
+++ b/classes/Http/Controller/TalkFormController.php
@@ -49,13 +49,11 @@ class TalkFormController extends BaseController
         // You can only edit talks while the CfP is open
         // This will redirect to "view" the talk in a read-only template
         if (!$this->callForPapers->isOpen()) {
-            $request->getSession()->set(
-                'flash',
-                [
-                    'type'  => 'error',
-                    'short' => 'Read Only',
-                    'ext'   => 'You cannot edit talks once the call for papers has ended', ]
-            );
+            $request->getSession()->set('flash', [
+                'type'  => 'error',
+                'short' => 'Read Only',
+                'ext'   => 'You cannot edit talks once the call for papers has ended',
+            ]);

             return new RedirectResponse($this->url('talk_view', ['id' => $talkId]));
         }
```

localheinz commented on classes/Http/Controller/TalkFormController.php:
```diff
@@ -0,0 +1,149 @@
+    {
+        // You can only create talks while the CfP is open
+        if (!$this->callForPapers->isOpen()) {
+            $request->getSession()->set(
```
See above.

localheinz commented on classes/Http/Controller/TalkFormController.php:
```diff
@@ -0,0 +1,149 @@
+            return $this->redirectTo('dashboard');
+        }
+
+        $data = [
```
While we are at it, can we inline `$data`?

Also see #851, where I already started working on that.

localheinz commented on classes/Http/Controller/TalkProcessingController.php:
```diff
@@ -0,0 +1,279 @@
+    private $mailer;
+    private $applicationEmail;
+    private $applicationTitle;
+    private $applicationEndDate;
```
Can we add docblocks here?

localheinz commented on classes/Http/Controller/TalkProcessingController.php:
```diff
@@ -0,0 +1,279 @@
+    {
+        // You can only create talks while the CfP is open
+        if (!$this->callForPapers->isOpen()) {
+            $request->getSession()->set(
```
See suggestion above in regard to wrapping.

localheinz commented on classes/Http/Controller/TalkProcessingController.php:
```diff
@@ -0,0 +1,279 @@
+
+        $user = $this->authentication->user();
+
+        $requestData = [
```
Most of the suggestions I make here for the *newly created* classes  are only because they are newly created classes, and as such, I think it's a good opportunity to clean up a few things. 🤓 

What do you think about inlining `$requestData` here? It's used once only.

localheinz commented on classes/Http/Controller/TalkProcessingController.php:
```diff
@@ -0,0 +1,279 @@
+                [
+                    'type'  => 'error',
+                    'short' => 'Read Only',
+                    'ext'   => 'You cannot edit talks once the call for papers has ended', ]
```
See suggestion made above in regard to wrapping.

localheinz commented on classes/Http/Controller/TalkProcessingController.php:
```diff
@@ -0,0 +1,279 @@
+     */
+    private function getTalkForm($requestData): TalkForm
+    {
+        $options = [
```
What do you think about inlining `$options` here?

BackEndTea commented on classes/Http/Controller/Admin/DashboardController.php:
```diff
@@ -20,16 +20,23 @@
 class DashboardController extends BaseController
 {
-    public function indexAction()
+    private $authentication;
```
Can we add a doc block here?

BackEndTea commented on classes/Http/Controller/Admin/ExportsController.php:
```diff
@@ -15,10 +15,21 @@
 
 class ExportsController extends BaseController
 {
+    private $session;
```
Can we add a doc block here?

BackEndTea commented on classes/Provider/Gateways/WebGatewayProvider.php:
```diff
@@ -32,6 +49,134 @@ class WebGatewayProvider implements BootableProviderInterface, ServiceProviderIn
+                $app['twig'],
+                $app['url_generator'],
+                $app->config('application.email'),
+                $app->config('application.title'),
```
I'm wondering if we can extract config to its own class, and inject that, rather than doing it this way.

Probably outside of the scope for this PR though.

localheinz replied:
@BackEndTea 

We need to be careful that we don't turn the configuration into just another container - maybe there's an opportunity to create a bunch of value objects instead. 👍

derrabus replied:
Yes, we could do that. However, in Symfony I'd convert these into parameters that would be injected via the DIC. I'm not a big friend on configuration objects, tbh.

chartjes replied:
@derrabus So how would that work? Please keep in mind I am not an experienced Symfony user, having only ever built this thing using Silex and never a more robust Symfony app.

chartjes replied:
@BackEndTea @localheinz I think decoupling the config object from being a property of an application object would be a good thing. Value objects containing these configuration values sound like a good idea to me.

derrabus replied:
@chartjes When building the container, you would parse your configuration file and add parameters to the container builder.
```php
$container->setParameter('application.title', $config['application']['title'];
```
This allows you to refer to that parameter in your service definition:
```yaml
services:
    OpenCFP\Http\Controller\TalkProcessingController:
    public: true
    autowire: true
    bind:
        $applicationEmail: '%application.email%'
        $applicationTitle: '%application.title%'
        $applicationEndDate: '%application.enddate%'
```
This way, the configuration is directly compiled into the container, so you don't have to parse the configuration file on every request.

mdwheele replied:
> Yes, we could do that. However, in Symfony I'd convert these into parameters that would be injected via the DIC. I'm not a big friend on configuration objects, tbh.

This is exactly what I've started doing.

BackEndTea commented on classes/Provider/Gateways/WebGatewayProvider.php:
```diff
@@ -32,6 +49,134 @@ class WebGatewayProvider implements BootableProviderInterface, ServiceProviderIn
     public function register(Container $app)
     {
+        // Controllers
+        $app[DashboardController::class] = function ($app) {
```
This function is becoming extremely large now.

Maybe we can extract the creation of the controllers to another provider.

derrabus replied:
If you migrate to Symfony, that class will be gone soon, anyway. I wouldn't put too much effort in splitting it up right now.

chartjes replied:
@derrabus There is no "if", we are definitely moving to Symfony 4!

derrabus commented:
I've squashed all commits and rebased against master to resolve some conflicts.

localheinz commented on tests/Integration/Http/Controller/DashboardControllerTest.php:
```diff
@@ -52,13 +52,13 @@ public function indexDisplaysUserAndTalks()
-        $profile->shouldReceive('name')->andReturn('Test User');
-        $profile->shouldReceive('photo', 'company', 'twitter', 'url', 'airport', 'bio', 'info', 'transportation', 'hotel');
-        $profile->shouldReceive('talks')->andReturn([$talk]);
+        $profile = m::mock(\OpenCFP\Domain\Speaker\SpeakerProfile::class);
```
👍

mdwheele commented:
I'm :+1: to merge after I get a change to review more in-depth. I have local changes towards migrating to Symfony 4 that includes parts of this diff. I think this is fine, but the amount of sweeping changes made in the same effort are making it more difficult to make progress, personally. That said, I'm also down to stop what I'm working on and try to organize the effort here to have it be less of a free-for-all. 

Thoughts?

localheinz commented:
@BackEndTea @chartjes @derrabus @mdwheele 

While this is great, if we feel this is too big of a change, maybe we can go with a hybrid approach:

1. adjust the [`ControllerResolver`](https://github.com/opencfp/opencfp/blob/30861d065293db51ed8c0365700d0d6bd9f4343e/classes/Provider/ControllerResolver.php) to first peek into the application container and fetch the controller from there, otherwise create it as usually
2. migrate one controller at a time, while at the same time decoupling it from the [`BaseController`](https://github.com/opencfp/opencfp/blob/30861d065293db51ed8c0365700d0d6bd9f4343e/classes/Http/Controller/BaseController.php) (which I think should eventually be removed
3. once done, remove the `BaseController` and the [`ContainerAware`](https://github.com/opencfp/opencfp/blob/30861d065293db51ed8c0365700d0d6bd9f4343e/classes/ContainerAware.php) trait

This would have the following advantage:

* work can be splitted into smaller steps
* smaller pull requests can be more easily reviewed
* we can be confident that the application still works even without significant test coverage as we are taking smaller steps

What do you think?

chartjes commented:
@localheinz I don't mind doing it all in one big shot like this PR. I do not disagree with your plan if I am outvoted in favour of a lower-risk process.

derrabus commented:
> I first built this with Silex because I wanted to use a micro framework out of the belief I could just wire stuff together. It should've been a Symfony (or other non-microframework) app from the start, and now we get to feel the pain caused by my early decisions.

@chartjes: Picking Silex was not a bad decision. Also, the codebase is really well maintained, so the migration path to Symfony isn't that much of a pain. Apart from the controllers, you have basically no direct dependencies on Silex. That's really great.

@localheinz: That's more or less how I developed this PR. Because I've squashed the commits for the last rebase, you don't see that progress anymore. Maybe I should've kept it. But anyway, since I haven changed any application logic, the diff is pretty straightforward. 

The only exception is the `TalkController` class that I've split. If you want me to, I merge it back into a single class with lots of dependencies and you can split it yourself later on.

localheinz commented:
@chartjes 

Since test coverage is low, smaller steps definitely carry lower risk! 👍

derrabus commented:
Okay, here's what I could offer if that helps you reviewing the code:

I submit my original first commit as a PR, with the following features:
* `BaseController` receives Twig and the URL generator as constructor arguments.
* All controllers are registered as services with those two dependencies.
* `ContainerAware` is removed from `BaseController` and added to each child controller.
* `ControllerResolver` is changed so it loads controllers from the service.

This PR should be easy to review. If you merge it to master, I submit one PR per controller that can be reviewed and merged individually. Also, I won't do any class splits this time. Once all of them are merged, we would have the same result as with this PR.

derrabus commented:
All right, I've submitted #855 according to my suggestion in the previous comment. If I get a quick merge on that one, I will prepare the individual PRs for the controllers.

derrabus commented:
Closing this PR, as it has been superseded by #855 and follow-ups.
